### PR TITLE
Handle cross-tab logout

### PR DESCRIPTION
**Pull request by adillpickles**
## Description
Adds cross-tab logout handling for BudgetBuddy by syncing the auth provider with changes to the `budgetbuddy.session` localStorage key. When a user logs out in one tab, other open tabs now re-read session state, leave authenticated UI, and rely on the existing protected layout redirect to send the user to `/login`.

## Related User Story / Issue
Fixes #100

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Backend / API change
- [ ] UI change
- [ ] Database change
- [ ] Documentation
- [ ] Refactor

## How Has This Been Tested?
- [X] GitHub Actions / CI tests passed
- [x] Manual testing performed
- [ ] Not tested locally

### Test Notes
Verified locally with:

- `npm test -- --runInBand __tests__/frontend/session.unit.test.js __tests__/frontend/providers.unit.test.js __tests__/frontend/appLayout.unit.test.js`
- `npm test -- --runInBand __tests__/frontend/dashboard-view.test.js __tests__/frontend/transactions.test.js __tests__/frontend/planner.test.js`
- `npm run build`
- `git diff --check`

Added focused provider tests for:
- responding to cross-tab removal of `budgetbuddy.session`
- ignoring unrelated storage events
- preserving immediate same-tab logout behavior

## UI Changes (if applicable)
- [x] No UI changes
- [ ] UI updated (add screenshots if needed)

## Notes for Reviewers
This is intentionally frontend-only and does not change backend auth/session behavior. The auth provider listens for the browser `storage` event on `budgetbuddy.session` and re-reads the session when that key changes. Routing remains handled by the existing protected app layout, and existing 401 expired-session handling is unchanged.

`BroadcastChannel` was not added because `localStorage` already provides the needed cross-tab event for this logout flow.

## Checklist
- [x] Linked to a user story or issue
- [x] Code builds / tests pass in CI
- [x] Ready for review

### CODE REVIEW

File: nextjs/__tests__/frontend/lib.session.unit.test.js
Comment (zakariac22) [inline nextjs/__tests__/frontend/lib.session.unit.test.js:1]: Let's first add cases that are supposed to break such as malformed JSON, to transitioning to valid JSON.  I would also recommend putting "lib." as a prefix for naming consistency

File: nextjs/__tests__/frontend/providers.unit.test.js
Comment (zakariac22) [inline nextjs/__tests__/frontend/providers.unit.test.js:1]: The tests are close to approval and demonstrate using the TDD approach for the main cases. To add to the approach, I would further suggest adding edge cases like missing stored sessions and malformed JSON inputs.

File: nextjs/__tests__/frontend/lib.session.unit.test.js
Comment (adillpickles) [reply nextjs/__tests__/frontend/lib.session.unit.test.js:1]: The edge cases were valid test gaps rather than runtime bugs, readSession already cleared malformed/invalid stored sessions and returned null for missing storage. I added tests for malformed JSON, missing/invalid stored sessions, and malformed to valid session recovery.

File: nextjs/__tests__/frontend/providers.unit.test.js
Comment (adillpickles) [reply nextjs/__tests__/frontend/providers.unit.test.js:1]: I added provider storage event coverage for missing session, malformed JSON, and malformed to valid cross tab transitions, plus kept same tab logout covered. Focused coverage now has session.js at 100% and providers.js above 90%. The same browser/same origin non incognito limitation is expected browser behavior.

File: nextjs/__tests__/frontend/providers.unit.test.js [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        oldValue: storedSession,
+        newValue: null,
+      }))
+    })
+
+    await waitFor(() => {
+      expect(screen.getByTestId('auth-state').textContent).toBe('signed-out')
+      expect(screen.getByTestId('token').textContent).toBe('none')
+    })
+  })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/frontend/providers.unit.test.js:100]: good catch, added a test for this. the old tests covered logout/malformed sessions, but not swapping from one valid session to another.
now it starts with one valid session, simulates another tab writing a new valid session, fires the storage event, and checks that the current tab updates to the new token/user. no runtime code changes, tests/build pass.

File: nextjs/src/components/providers.js [hunk 2]
```diff
@@ -63,6 +63,16 @@ function AuthProvider({ children }) {
...
 
+  useEffect(() => {
+    const handleStorage = (event) => {
+      if (event.key !== SESSION_STORAGE_KEY) return
+      setSession(readSession())
+    }
+
+    window.addEventListener('storage', handleStorage)
+    return () => window.removeEventListener('storage', handleStorage)
+  }, [])
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/providers.js:76]: good catch, this was valid. I updated the auth storage listener to treat key: null as a session-affecting storage event from localStorage.clear(), added a focused provider test, and confirmed targeted Jest, full Jest, and build all pass.

File: nextjs/__tests__/frontend/providers.unit.test.js
Comment (b-Karthikeya-reddy) [inline nextjs/__tests__/frontend/providers.unit.test.js:1]: Small test cleanup: in ignores unrelated storage events, consider dropping localStorage.removeItem(SESSION_STORAGE_KEY) unless you explicitly want layered behavior. Keeping the stored session intact makes the test read as “unrelated keys don’t affect auth,” which matches what you dispatch.

File: nextjs/src/components/providers.js
Comment (b-Karthikeya-reddy) [inline nextjs/src/components/providers.js:1]: Consider also checking event.storageArea === window.localStorage so you never accidentally react to sessionStorage storage events if key/shape ever overlaps in theory

**Review by b-Karthikeya-reddy — CHANGES_REQUESTED** [review]

File: nextjs/__tests__/frontend/providers.unit.test.js
Comment (adillpickles) [reply nextjs/__tests__/frontend/providers.unit.test.js:1]: simplified this so the stored session stays intact while the unrelated localStorage event fires, and the assertion now directly confirms auth remains signed in.

File: nextjs/src/components/providers.js
Comment (adillpickles) [reply nextjs/src/components/providers.js:1]: added the localStorage storageArea guard and covered the sessionStorage overlap case, while keeping the existing localStorage session-key and key: null clear behavior intact.

**Review by zakariac22 — APPROVED** [review]
The necessary changes were made

**Review by b-Karthikeya-reddy — APPROVED** [review]
looks good to be merged
